File: app/src/main/java/com/absinthe/anywhere_/BaseActivity.java
package com.absinthe.anywhere_;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;

import com.absinthe.anywhere_.interfaces.OnDocumentResultListener;
import com.absinthe.anywhere_.model.Const;
import com.absinthe.anywhere_.model.GlobalValues;
import com.absinthe.anywhere_.ui.main.MainActivity;
import com.absinthe.anywhere_.utils.AppUtils;
import com.absinthe.anywhere_.utils.StatusBarUtil;
import com.absinthe.anywhere_.utils.UiUtils;
import com.absinthe.anywhere_.utils.manager.Logger;

@SuppressLint("Registered")
public abstract class BaseActivity extends AppCompatActivity {

    protected Toolbar mToolbar;
    private OnDocumentResultListener mListener;

    protected abstract void setViewBinding();
    protected abstract void setToolbar();
    protected abstract boolean isPaddingToolbar();

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Logger.i(this.getClass().getSimpleName(), "onCreate");
        setViewBinding();
        initView();
    }

    protected void initView() {
        if (GlobalValues.sBackgroundUri.isEmpty() || !(this instanceof MainActivity)) {
            if (UiUtils.isDarkMode(this)) {
                StatusBarUtil.setDarkMode(this);
            } else {
                StatusBarUtil.setLightMode(this);
            }
        }

        setToolbar();
        if (isPaddingToolbar()) {
            mToolbar.setPadding(0, StatusBarUtil.getStatusBarHeight(this), 0, 0);
        }
        setSupportActionBar(mToolbar);
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
        }
    }

    public void restartActivity() {
        Intent intent = getIntent();
        finish();
        startActivity(intent);
    }

    public void setDocumentResultListener(OnDocumentResultListener listener) {
        mListener = listener;
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        if (requestCode == Const.REQUEST_CODE_IMAGE_CAPTURE && resultCode == RESULT_OK) {
            if (mListener != null && data != null && data.getData() != null) {
                mListener.onResult(data.getData());
                AppUtils.takePersistableUriPermission(this, data.getData(), data);
            }
        }
        super.onActivityResult(requestCode, resultCode, data);
    }

    @Override
    public void finish() {
        if (GlobalValues.sIsExcludeFromRecent) {
            finishAndRemoveTask();
        } else {
            super.finish();
        }
    }
}

File: app/src/main/java/com/absinthe/anywhere_/BaseActivity.kt
package com.absinthe.anywhere_

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.res.Configuration
import android.content.res.Resources
import android.graphics.Color
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.os.FileUriExposedException
import android.view.MenuItem
import android.view.View
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContract
import androidx.annotation.CallSuper
import androidx.appcompat.app.AppCompatActivity
import androidx.viewbinding.ViewBinding
import com.absinthe.anywhere_.constants.GlobalValues
import com.absinthe.anywhere_.utils.AppUtils
import com.absinthe.anywhere_.utils.ToastUtil
import com.absinthe.anywhere_.utils.manager.ActivityStackManager
import rikka.core.res.isNight
import rikka.material.app.MaterialActivity
import timber.log.Timber
import java.lang.ref.WeakReference


@SuppressLint("Registered, MissingSuperCall")
abstract class BaseActivity<T : ViewBinding> : MaterialActivity() {

  var shouldFinishOnResume = false

  private var onDocumentResultAction: ((uri: Uri) -> Unit)? = null
  private lateinit var reference: WeakReference<AppCompatActivity>
  private lateinit var openDocumentResultLauncher: ActivityResultLauncher<Array<String>>

  protected lateinit var binding: T
  protected lateinit var root: View

  protected abstract fun setViewBinding(): T?

  override fun onCreate(savedInstanceState: Bundle?) {
    Timber.i("onCreate")
    super.onCreate(savedInstanceState)

    reference = WeakReference(this)
    ActivityStackManager.addActivity(reference)

    setViewBinding()?.let {
      binding = it
      root = binding.root
      setContentView(root)
    }
    initView()

    openDocumentResultLauncher = registerForActivityResult(OpenDocument()) {
      try {
        it?.let {
          it.data?.let { uri ->
            onDocumentResultAction?.invoke(uri)
            if (uri.toString().contains("file://")) {
              ToastUtil.makeText(R.string.toast_file_uri_exposed)
            } else {
              try {
                AppUtils.takePersistableUriPermission(this, uri, it)
              } catch (e: RuntimeException) {
                ToastUtil.makeText(R.string.toast_runtime_error)
              }
            }
          }
        }
      } catch (e: Exception) {
        if (AppUtils.atLeastN()) {
          if (e is FileUriExposedException) {
            ToastUtil.makeText(R.string.toast_file_uri_exposed)
          }
        }
      }
    }
  }

  override fun onResume() {
    super.onResume()
    if (shouldFinishOnResume) {
      finish()
    }
  }

  override fun onDestroy() {
    ActivityStackManager.removeActivity(reference)
    super.onDestroy()
  }

  override fun shouldApplyTranslucentSystemBars(): Boolean {
    return true
  }

  override fun computeUserThemeKey(): String {
    return GlobalValues.darkMode
  }

  override fun onApplyUserThemeResource(theme: Resources.Theme, isDecorView: Boolean) {
    if (AppUtils.atLeastS()) {
      if (resources.configuration.isNight()) {
        theme.applyStyle(R.style.ThemeOverlay_DynamicColors_Dark, true)
      } else {
        theme.applyStyle(R.style.ThemeOverlay_DynamicColors_Light, true)
      }
    }
    theme.applyStyle(R.style.ThemeOverlay, true)
  }

  override fun onApplyTranslucentSystemBars() {
    super.onApplyTranslucentSystemBars()
    window.statusBarColor = Color.TRANSPARENT
    window.decorView.post {
      window.navigationBarColor = Color.TRANSPARENT
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
        window.isNavigationBarContrastEnforced = false
      }
    }
  }

  fun setDocumentResult(mimeType: String, action: ((uri: Uri) -> Unit)?) {
    onDocumentResultAction = action
    openDocumentResultLauncher.launch(arrayOf(mimeType))
  }

  override fun onOptionsItemSelected(item: MenuItem): Boolean {
    if (item.itemId == android.R.id.home) {
      onBackPressed()
    }
    return super.onOptionsItemSelected(item)
  }

  override fun finish() {
    if (GlobalValues.isExcludeFromRecent) {
      finishAndRemoveTask()
    } else {
      super.finish()
    }
  }

  protected open fun initView() {}

  fun isNightMode(): Boolean {
    return resources.configuration.uiMode and Configuration.UI_MODE_NIGHT_YES > 0
  }

  class OpenDocument : ActivityResultContract<Array<String>, Intent?>() {
    @CallSuper
    override fun createIntent(context: Context, input: Array<String>): Intent {
      return Intent(Intent.ACTION_OPEN_DOCUMENT)
        .putExtra(Intent.EXTRA_MIME_TYPES, input)
        .setType("*/*")
    }

    override fun getSynchronousResult(
      context: Context,
      input: Array<String>
    ): SynchronousResult<Intent?>? = null

    override fun parseResult(resultCode: Int, intent: Intent?): Intent? {
      return intent.takeIf { resultCode == Activity.RESULT_OK }
    }
  }
}
